### DISCUSSION TITLE
Converting a Razor pages application with using OrchardCore and Razor Pages in conjuction

### DISCUSSION THREAD

**Discussion by rwxzig** [OrchardCMS/OrchardCore#14285] [Q&A] [answered]
I am new to OrchardCore and I love it so far, after a huge research for which CMS would suit my needs best, I decided to go with Orchard, but I need to learn how to do certain things according to my needs. 
I have an app structured as this:
```
Pages
--Home
----Index.cshtml
----Index.cshtml.cs
--About
etc
```
Inside  `Index.cshtml.cs` I have my HomeModel and its constructor, plus endpoints OnGet, OnPost, OnGetDataAsync, OnPostVehicleData etc...

Here is an example of Index.cshtml.cs
```cs
namespace Demo.Pages;

public class HomeIndexModel : PageModel
{
  private readonly ILogService _logService;
  public HomeIndexModel(ILogService logService)
  {
    _logService = logService;
  }
  public async Task<IActionResult> OnGet()
  {
    // .... Do something
    return Page();
  }

  public async Task<IActionResult> OnGetDataAsync(string value)
  {
    // Do something else
    // var json = ....
    return Content(json);
  }
  public async Task<IActionResult> OnPostVehicleData ([FromBody] InputModel data) 
  {
    // Do something
    // var json = await _someService.PostVehicleDataAt(data);
    return Content(json);
  }
}
```
And this is the function that calls the endpoint in `Index.cshtml`
```js
function callSomeApiEndpoint () {
	var data = { "item": 1, id: 2}
		$.ajax({
				type: 'POST',
				beforeSend: function (xhr) {
                    xhr.setRequestHeader("XSRF-TOKEN",
                        $('input:hidden[name="__RequestVerificationToken"]').val());
                },
				url: '@Url.Page("HomeIndex")?handler=VehicleData',
				data: data,
				success: function (data) {
					var res = JSON.parse(data);
					// Do something
				},
				error: function (err) {
					// Toast some info
				}
		});
}
```

The OnGet method is called with a GET request, but when I try to call other endpoints, it just does not find them. 

I think it has to do with the way how MapRazorPages() works, and I was looking for a solution since I cannot add `app.MapRazorPages();`

My question is how can I achieve the same goal using OrchardCore, after searching in internet I think a solution would be to create a Module, but I did not find much info on how to implement that.
Second question, is it possible to add controllers?

**Comment by Piedone** [comment, marked as answer]
Awesome that you found Orchard!

I assume you've seen the [decoupled docs page](https://docs.orchardcore.net/en/latest/docs/guides/decoupled-cms/); if not, I suggest you start there.

You don't need to call `MapRazorPages()` because that already happens within Orchard. Otherwise, Razor Pages should work exactly the same how they do outside of an Orchard Core-using app.

What happens if you just try `OnPost()`?

Yes, you can also add controllers. For Razor Pages, usually you'd add all files into the root web project (under the Pages folder/namespace), and you can similarly add MVC controllers to it. You can also do that in a module indeed. Adding a `Manifest` file and a package reference to `OrchardCore.Module.Targets` will make a project into a module, what then you can enable from the Orchard admin. And in it, you can do everything you know from ASP.NET Core MVC, plus what Orchard provides. For more info, I suggest you check out the https://docs.orchardcore.net/en/latest/docs/getting-started/templates/ and https://docs.orchardcore.net/en/latest/docs/getting-started/templates/ docs pages.

**Reply by rwxzig** [reply]
Yes I have read the docs and re reading again, because I am sure that this feature can be implemented, just don't know yet how. 
[This](https://github.com/OrchardCMS/OrchardCore/issues/3449) lead me to Modules, and I tried to follow the steps and enable it to admin, but I just could not get it to make a POST request to the endpoint that I created, I get the classic "The page could not be found". 

If anyone else had the same problem, would love to hear how the implemented this, or if they have some practical example then I would love to see.

**Reply by rwxzig** [reply]
@Piedone thanks for your reply and keep up the good work.

The way how I solved this, is by creating a Controller folder, and I needed to add `.AddMvc()` on `builder.Services.AddOrchardCms();` 
There are perhaps better way on doing this, but for now it does the job so I will keep it as it, but when I learn more about Orchard I will probably need to do a huge refactoring.

**Reply by Piedone** [reply]
Great then!

What I mean is that for Razor Pages you don't need to create modules; you can put all your code into the root web project. (I don't actually know if they'd still work in a module, unlike MVC controllers and views.)

You shouldn't need to use `AddMvc()` either since that's also done by Orchard (in `AddOrchardCms()`). 

But if it works then probably it's the best to continue instead of debugging it :).

**Comment by hishamco** [comment]
Maybe I came late, but I did a lot of work with OC & RazorPages, so my question why do you need JavaScript? If your answer is AJAX, then why do you need the `PageModel`?

I usually use the page model to interact with OC APIs (OC.ContentManagement, OC.Media ... etc)

> MapRazorPages()

No need to call it as @Piedone mentioned before

**Reply by rwxzig** [reply]
I took some example code, we have legacy code that I do not want to rewrite it all so I am trying to learn so I can do minimal changes without breaking any functionality. We use `DataGrid` for displaying all sorts of data, and some operations are done in the page model.
But of course there might be better ways of implementing it.

**Comment by ns8482e** [comment]
Looks like you need to define route with handler

`@page "/{handler}"`

Also check demo module, to learn more about razor pages route mapping

`src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs`

**Reply by hishamco** [reply]
> @page "/{handler}"

`@page` is required of course
